File: AsianOptionsServiceFSharpLibCall/AsianOptionsServiceFSharp.svc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.ServiceModel;
using System.ServiceModel.Web;
using System.Text;
using AOSF = AsianOptionsServiceFSharpLib;

namespace AsianOptionsServiceFSharpLibCall
{
    [ServiceBehavior(ConcurrencyMode = ConcurrencyMode.Multiple)]
    public class AsianOptionsServiceFSharp : IAsianOptionsServiceFSharp
    {
        /// <summary>
        /// Calculates the price of an option with the given 
        /// statistical information. The algorithm used for calculation 
        /// is based on the Montecarlo method.
        /// </summary>
        /// <param name="initial">Intial value.</param>
        /// <param name="exercise">Excercise value.</param>
        /// <param name="up">Up value.</param>
        /// <param name="down">Down value.</param>
        /// <param name="interest">Intrest value.</param>
        /// <param name="periods">Number of periods value.</param>
        /// <param name="runs">Number of runs.</param>
        /// <returns>The calculated value for an option with the given 
        /// statistical context using the Montecarlo Method.</returns>
        public double PriceAsianOptions(double initial, double exercise, double up, double down, double interest, int periods, int runs)
        {
            string x = AOSF.Class1.X();
            double[] pricePath = new double[periods + 1];

            // Risk-neutral probabilities
            double piup = (interest - down) / (up - down);
            double pidown = 1 - piup;

            double temp = 0.0;

            Random rand = new Random();
            double priceAverage = 0.0;
            double callPayOff = 0.0;

            for (int index = 0; index < runs; index++)
            {
                // Generate Path
                double sumPricePath = initial;

                for (int i = 1; i <= periods; i++)
                {
                    pricePath[0] = initial;
                    double rn = rand.NextDouble();

                    if (rn > pidown)
                    {
                        pricePath[i] = pricePath[i - 1] * up;
                    }
                    else
                    {
                        pricePath[i] = pricePath[i - 1] * down;
                    }

                    sumPricePath += pricePath[i];
                }

                priceAverage = sumPricePath / (periods + 1);
                callPayOff = Math.Max(priceAverage - exercise, 0);

                temp += callPayOff;
            }

            return (temp / Math.Pow(interest, periods)) / runs;
        }

    }
}

File: AsianOptionsServiceFSharp/Services/AsianOptionsServiceFSharp.svc.fs
namespace AsianOptionsServiceFSharpLib.Services

open System
open AsianOptionsServiceFSharpLib.Contracts

type AsianOptionsFSharp() =
    interface IAsianOptionsServiceFSharp with
        override this.PriceAsianOptions(initial:float, exercise:float, up:float
            , down:float, interest:float, periods:int32, runs:int32) = 
            
            (*let initial = 30.0;;
            let exercise = 30.0;;
            let up = 1.4;;
            let down = 0.8;;
            let interest = 1.08;;
            let periods = 20;;
            let runs = 1000000;;*)

            let pricePath = Array.create (periods + 1) 0.0
            let piup = (interest - down) / (up - down)
            let pidown = ref (1.0 - piup)
            let temp = ref 0.0
            let rand = new System.Random()
            let priceAverage = 0.0
            let callPayOff = 0.0
            {0..(runs - 1)} |> Seq.iter (
                fun _ ->
                    let sumPricePath = ref 0.0
                    sumPricePath.Value <- initial  
                    
                    {1..periods} |> Seq.iter (
                        fun i -> 
                            pricePath.[0] <- initial
                            let rn = rand.NextDouble()
                            match rn with
                            | x when x > pidown.Value ->
                                pricePath.[i] <- pricePath.[i - 1] * up
                            | _ ->
                                pricePath.[i] <- pricePath.[i - 1] * down
                            sumPricePath.Value <- sumPricePath.Value + pricePath.[i] //*)
                    );
                    let priceAverage = sumPricePath.Value / (float periods + 1.0) * 1.0
                    let callPayOff = System.Math.Max((priceAverage - exercise), 0.0)
                    temp.Value <- temp.Value + callPayOff
            )

            (temp.Value / System.Math.Pow(interest, float periods)) / (float runs);
            //*)